src/effects/VST/VSTEffect.cpp

```diff
@@ -3463,4 +3463,94 @@ void VSTEffectWrapper::ForEachParameter(ParameterVisitor visitor) const
 }
 
 
+bool VSTEffectWrapper::FetchSettings(VSTEffectSettings& vst3settings) const
+{
+   // Get the fallback ID-value parameters
+   ForEachParameter
+   (
+      [&](const ParameterInfo& pi)
+      {
+         float val = callGetParameter(pi.mID);
+         vst3settings.mParamsMap[pi.mName] = val;
+         return true;
+      }
+   );
+
+   // These are here to be checked against for compatibility later
+   vst3settings.mVersion   = mAEffect->version;
+   vst3settings.mUniqueID  = mAEffect->uniqueID;
+   vst3settings.mNumParams = mAEffect->numParams;
+
+   // Get the chunk (if supported)
+   vst3settings.mChunk = std::nullopt;
+   if (mAEffect->flags & effFlagsProgramChunks)
+   {
+      void* chunk = NULL;
+      int clen = (int)constCallDispatcher(effGetChunk, 1, 0, &chunk, 0.0);
+      if (clen > 0)
+      {
+         vst3settings.mChunk = Base64::Encode(chunk, clen);
+      }
+   }
+
+   return true;
+}
+
+
+bool VSTEffectWrapper::StoreSettings(const VSTEffectSettings& vst3settings)
+{
+   // First, make sure settings are compatibile with the plugin
+   if ((vst3settings.mUniqueID  != mAEffect->uniqueID)   ||
+       (vst3settings.mVersion   != mAEffect->version)    ||
+       (vst3settings.mNumParams != mAEffect->numParams)      )
+   {
+      return false;
+   }
+
+
+   // Try using the chunk first (if available)
+   if (vst3settings.mChunk)
+   {
+      ArrayOf<char> buf{ vst3settings.mChunk->length() / 4 * 3 };
+
+      int len = Base64::Decode(*vst3settings.mChunk, buf.get());
+      if (len)
+      {
+         VstPatchChunkInfo info = { 1, mAEffect->uniqueID, mAEffect->version, mAEffect->numParams, "" };
+
+         callSetChunkB(true, len, buf.get(), &info);
+         return true;
+      }         
+   }
+
+
+   // Chunk unavailable / decoding it did not work? fall back to param ID-value pairs
+   const auto vstVersion = constCallDispatcher(effGetVstVersion, 0, 0, NULL, 0);
+
+   ForEachParameter
```

Comment:
This visit of parameters has these differences from `VSTEffect::LoadSettings`:
- The bracketing dispatches of `effBeginSetProgram` and `effEndSetProgram` are not done
- The bounds check on `value` between +/- 1 are not done

Is it important to preserve the bounds check?  I honestly don't know.  Preserve it, as real code or as commented out, and experiment may later decide if we really need it.

`LoadSettings` as I now see, and as you probably saw too, was doing `callSetParameter` on self and then again on slaves:  which was unnecessary, because the `callSetParameter` call on self was already recursive.

The begin and end of set program were only done on master.  Was it right or wrong not to do it to on slaves?  I don't know.  But let's assume, it can be done on every handle for every instance, and the calls should therefore be part `StoreSettings.`

In practice, multiple slaves will happen only when you use the old non-modal dialog from the Effects menu, which will eventually not be part of 3.2 anyway.  But we may in future want to reimplement something like it for a master bus of effects.

Reply:
I added the bounds checking and the  `effBeginSetProgram` / `effEndSetProgram` bracketing. I need a plugin which does not support chunks to test this, but since LoadSettings did this, it looks ok if StoreSettings also does it.